tests/rfxencode.c

```diff
@@ -0,0 +1,325 @@
+    unsigned int   image_size;
+    int            hres;
+    int            vres;
+    unsigned int   ncolors;
+    unsigned int   nimpcolors;
+};
+
+static int
+read_bitmap(char *file_name, int *width, int *height, int *bpp, char *bmp_data)
+{
+    int i;
+    int j;
+    int e;
+    int fd;
+    int file_stride_bytes;
+    int pixel;
+    struct bmp_magic bm;
+    struct bmp_hdr bh;
+    struct dib_hdr dh;
+    unsigned char* src8;
+    int* src32;
+    int* dst32;
+
+    fd = open(file_name, O_RDONLY);
+    if (fd == -1)
+    {
+        return 1;
+    }
+    if (read(fd, &bm, sizeof(bm)) != sizeof(bm))
+    {
+        close(fd);
+        return 1;
+    }
+    if (read(fd, &bh, sizeof(bh)) != sizeof(bh))
+    {
+        close(fd);
+        return 1;
+    }
+    if (read(fd, &dh, sizeof(dh)) != sizeof(dh))
+    {
+        close(fd);
+        return 1;
+    }
+
+    if (dh.bpp != 24 && dh.bpp != 32)
+    {
+        printf("error, only support 24 and 32 bpp\n");
+        close(fd);
+        return 1;
+    }
+
+    *width = dh.width;
+    *height = dh.height;
+    *bpp = dh.bpp;
+
+    file_stride_bytes = dh.width * ((dh.bpp + 7) / 8);
+    e = (4 - file_stride_bytes) & 3;
+    src8 = (unsigned char*)malloc(file_stride_bytes * 4);
+    for (j = 0; j < dh.height; j++)
+    {
+        dst32 = (int*)(bmp_data + (dh.width * dh.height * 4) - ((j + 1) * dh.width * 4));
+        if (read(fd, src8, file_stride_bytes + e) != file_stride_bytes + e)
+        {
+            free(src8);
+            close(fd);
+            return 1;
+        }
+        if (dh.bpp == 32)
+        {
+            src32 = (int*)src8;
+            for (i = 0; i < dh.width; i++)
+            {
+                pixel = src32[i];
+                dst32[i] = pixel;
+            }
+        }
+        else if (dh.bpp == 24)
+        {
+            for (i = 0; i < dh.width; i++)
+            {
+                pixel =  src8[i * 3 + 0] << 0;
+                pixel |= src8[i * 3 + 1] << 8;
+                pixel |= src8[i * 3 + 2] << 16;
+                dst32[i] = pixel;
+            }
+        }
+    }
+
+    free(src8);
+    close(fd); 
+    return 0;
+}
+
+int out_params(void)
+{
+    printf("rfxencode: a RemoteFX encoder testing program.\n");
+    printf("  -i <in file name> bmp file\n");
+    printf("  -o <out file name> rfx file\n");
+    printf("  -c <number> times to loop\n");
+    printf("  -n no accel\n");
+    printf("  -1 use rlgr1\n");
+    return 0;
+}
+
+int process(void)
+{
+    char *out_data;
+    char *bmp_data;
+    int out_fd;
+    int out_bytes;
+    int error;
+    int index;
+    int jndex;
+    void* han;
+    int num_tiles;
+    int num_tiles_x;
+    int num_tiles_y;
+    int flags;
+    int width;
+    int height;
+    int bpp;
+    struct rfx_rect region;
+    struct rfx_tile *tiles;
+    struct rfx_tile *tile;
+
+    out_data = (char *)malloc(1024 * 1024);
+    bmp_data = (char *)malloc(10 * 1024 * 1024);
+    memset(bmp_data, 0xff, 10 * 1024 * 1024);
```

Comment:
I would prefer to has as few magic numbers as possible. Let's use named constants.